### PR TITLE
Rebuild reminders notifications

### PR DESCRIPTION
**Pull request by amrhossamdev**
Fixes: #1258

<p align="center">
    alarms
</p>

## Key Changes

- I removed the old implementation of trip service alerts.
- Integrated the OneSignal SDK with our backend services to create alarms.
- Migrated trips db and updated db version

## How Reminders Work

- The user selects a stop and clicks on the bus arrival times to set a reminder.
- A page opens where the user can specify when they want to receive the alarm.
- A POST request is sent to the admin server to create the alarm, which is also saved locally.
- Once the user receives the bus arrival alarm, the alarm is deleted from the local database.
- If the user edits an alarm, a DELETE request is sent to remove the existing alarm, followed by a new request to create the updated one.
- If the user clicks on the alarm I will open the `TripInfoActivity` that contains all arrivals 
- When a user receives a notification the alarm will be deleted from the local db.

## Tasks 
- [x] Apply the `AndroidStyle.xml` style template to your code in Android Studio.

- [x] Run the unit tests with `gradlew connectedObaGoogleDebugAndroidTest` to make sure you didn't break anything

- [x] If you have multiple commits please combine them into one commit by squashing them for the initial submission of the pull request.  When addressing comments on a pull request, please push a new commit per comment when possible (reviewers will squash and merge using GitHub merge tool)

### CODE REVIEW

File: BUILD.md [hunk 1]
```diff
@@ -50,6 +50,16 @@ Add the following to `onebusaway-android/gradle.properties`:
 
 ...where `XXXXXX` is your API key. Note that the suffix of `_oba` can be changed to configure API keys for other build flavors.
 
+### Configuration for OneSignal App ID
+
+If push arrivals notifications are active, you'll need to provide an App ID for [OneSignal](https://onesignal.com/).
+
+Add the following to `onebusaway-android/gradle.properties`:
```
Comment (aaronbrethorst) [inline BUILD.md:57]: I see you have `ONESIGNAL_APP_ID="your_onesignal_app_id"` in gradle.properties in the project root. Which gradle.properties should this be set in, and can you clarify whether the key should be `ONESIGNAL_APP_ID` or `OneSignal_App_ID`?

File: onebusaway-android/src/main/java/org/onebusaway/android/io/request/reminders/ObaReminderRequest.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+    private final ReminderRequestListener listener;
+    private static final String TAG = "ObaReminderRequest";
+
+    private ObaReminderRequest(Uri uri, String body, ReminderRequestListener listener) {
+        super(uri, body);
+        this.listener = listener;
+    }
+
+    public static class Builder extends PostBuilderBase {
```
Comment (aaronbrethorst) [inline onebusaway-android/src/main/java/org/onebusaway/android/io/request/reminders/ObaReminderRequest.java:30]: I think you should add some validation of the values that are being included in the POST body, specifically verifying that all necessary values are included before the POST is sent. I'll leave it to you to determine the best location :)

**Review by aaronbrethorst — CHANGES_REQUESTED** [review]
The code is generally looking great! I'm running into some trouble getting the feature to work as expected, though:

1. Don't show reminder times that occur outside of the departure window. for example, a bus departing in 10 minutes, should not give me the option for scheduling a reminder that occurs for > 10 minutes (e.g. 15, 20, 30)

2. I'm not being prompted for push notification permissions when I try to create a reminder. Also, here's what I see in logcat:

```
2024-10-07 16:05:05.603 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  mStopID: 1_19790
2024-10-07 16:05:05.603 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  mRouteId: 1_102576
2024-10-07 16:05:05.603 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  mReminderTime: 180
2024-10-07 16:05:05.603 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  mStopSequence: 10
2024-10-07 16:05:05.603 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  mServiceDate: 1728284400000
2024-10-07 16:05:05.603 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  mTripID: 1_635649727
2024-10-07 16:05:05.603 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  mVehicleID: 1_6246
2024-10-07 16:05:05.604 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  user-push-id: 
2024-10-07 16:05:05.607 18248-18546 ObaDefaultConnection    com.joulespersecond.seattlebusbot    D  https://onebusaway.co/api/v1/regions/1/alarms
2024-10-07 16:05:05.864 18248-18546 ObaReminderRequest      com.joulespersecond.seattlebusbot    D  Response received: ReminderResponse{url='null'}
2024-10-07 16:05:05.865 18248-18248 TripInfoActivity        com.joulespersecond.seattlebusbot    D  Response or url is null
```

I think this might be because I forgot to set `OneSignal_App_ID`. (update: nope. It's still not showing the permission prompt after setting that value.)

3. I think the reminder feature should be hidden from the app if `OneSignal_App_ID` value is not set.

File: BUILD.md (hunk 1 above)
Comment (amrhossamdev) [reply BUILD.md:57]: ONESIGNAL_APP_ID

File: onebusaway-android/src/main/java/org/onebusaway/android/io/request/reminders/ObaReminderRequest.java (hunk 2 above)
Comment (amrhossamdev) [reply onebusaway-android/src/main/java/org/onebusaway/android/io/request/reminders/ObaReminderRequest.java:30]: Correct.

File: BUILD.md (hunk 1 above)
Comment (amrhossamdev) [reply BUILD.md:57]: I have updated the `build.md` with the correct configurations

File: onebusaway-android/src/main/java/org/onebusaway/android/ui/TripInfoActivity.java [hunk 3]
```diff
@@ -407,186 +433,161 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
-                ActivityCompat.requestPermissions(this.getActivity(),
-                        new String[] {Manifest.permission.POST_NOTIFICATIONS},
-                        NOTIFICATION_PERMISSION_REQUEST);
+        public void onSave() {
+            // If update delete alarm before creating a new one
+            if (ReminderUtils.isAlarmExist(getContext(), mTripUri)) {
+                requestDeleteAlarm(getContext(), mTripUri);
             }
-
+            OneSignal.getNotifications().requestPermission(false, Continue.none());
```
Comment (amrhossamdev) [inline onebusaway-android/src/main/java/org/onebusaway/android/ui/TripInfoActivity.java:447]: Prompting for receiving notification will be for Android version >=14

**Review by aaronbrethorst — APPROVED** [review]
terrific work 👍

**Comment by amrhossamdev** [comment]
> terrific work 👍 

🚀❤️❤️

**Pull request merged**